### PR TITLE
feat(etherscan directory output)

### PR DESCRIPTION
**Pull request by jubos**
* Inspired by etherscan.deth.net, this feature will expand a verified contract's
  source code paths to a directory for easy consumption with your favorite
  editor / IDE.

* Tested via the following examples that test a few of the edge cases.
> cargo run --bin cast -- etherscan-source 0x8d04a8c79cEB0889Bdd12acdF3Fa9D207eD3Ff63 -d blitmap
This results in a directory tree with all the source files of the contract.

> cargo run --bin cast -- etherscan-source 0xBB9bc244D798123fDe783fCc1C72d3Bb8C189413 -d thedao
This results in a single file output.

> cargo run --bin cast -- etherscan-source 0xb5c31a0e22cae98ac08233e512bd627885aa24e5 -d unverified
This returns an error since the following contract is not verified yet on Etherscan

## Motivation

When learning about smart contracts it is a nice improvement to able the browse the contract code in your normal developer environment.

## Solution

This feature parses the etherscan API source structure and writes it to directory.

### CODE REVIEW

File: cast/src/lib.rs [hunk 1]
```diff
@@ -1209,6 +1225,75 @@ impl SimpleCast {
...
+    /// ```
+    pub async fn expand_etherscan_source_to_directory(
+        chain: Chain,
+        contract_address: String,
+        etherscan_api_key: String,
+        output_directory: PathBuf,
+    ) -> eyre::Result<()> {
+        let client = Client::new(chain, etherscan_api_key)?;
+        let meta = client.contract_source_code(contract_address.parse()?).await?;
+        if meta.items.iter().any(|item| item.abi == "Contract source code not verified") {
```
Comment (mattsse) [inline cast/src/lib.rs:1272]: we should handle this while deserializing the response in the etherscan client and make this an Option. Mind opening an issue for this on ethers or even a PR?

File: cast/src/lib.rs [hunk 2]
```diff
@@ -1209,6 +1225,75 @@ impl SimpleCast {
...
+    /// Etherscan can either return one raw string that includes all of the solidity for a verified
+    /// contract or a json struct surrounded in an extra set of {} that includes a directory
+    /// structure with paths and source code.
+    fn get_sources_from_etherscan_source_value(
+        contract_name: &str,
+        etherscan_source: &str,
+    ) -> Result<Vec<(String, String)>> {
+        if etherscan_source.len() >= 4 &&
+            &etherscan_source[0..2] == "{{" &&
+            &etherscan_source[etherscan_source.len() - 2..] == "}}"
```
Comment (mattsse) [inline cast/src/lib.rs:1238]: use `.ends_with` and `starts_with` instead

File: cast/src/lib.rs [hunk 3]
```diff
@@ -1240,6 +1325,28 @@ fn strip_0x(s: &str) -> &str {
...
 }
 
+/// Ensure that input such as a smart contract name won't cause a directory traversal or has
+/// directory separators.
+fn sanitize_single_component(component: &str) -> String {
+    Path::new(component)
+        .components()
+        .filter_map(|x| {
+            let os_str = x.as_os_str();
+            if os_str == "." || os_str == ".." {
```
Comment (mattsse) [inline cast/src/lib.rs:1335]: you can use `x == Component::CurDir || x == Component::ParentDir` instead

File: cast/src/lib.rs [hunk 4]
```diff
@@ -1209,6 +1225,75 @@ impl SimpleCast {
...
+    fn get_sources_from_etherscan_source_value(
+        contract_name: &str,
+        etherscan_source: &str,
+    ) -> Result<Vec<(String, String)>> {
+        if etherscan_source.len() >= 4 &&
+            &etherscan_source[0..2] == "{{" &&
+            &etherscan_source[etherscan_source.len() - 2..] == "}}"
+        {
+            let json = &etherscan_source[1..etherscan_source.len() - 1];
+            let parsed: EtherscanSourceJsonMetadata = serde_json::from_str(&json)?;
```
Comment (mattsse) [inline cast/src/lib.rs:1241]: think this should be `json` instead of `&json`

File: cast/src/lib.rs [hunk 5]
```diff
@@ -1240,6 +1325,28 @@ fn strip_0x(s: &str) -> &str {
...
+                os_str.to_str()
+            }
+        })
+        .collect::<Vec<&str>>()
+        .join("-")
+}
+
+/// Remove any components in a smart contract source path that could cause a directory traversal.
+fn sanitize_path(path: &str) -> PathBuf {
+    Path::new(path).components().filter(|x| x.as_os_str() != "..").collect::<PathBuf>()
```
Comment (mattsse) [inline cast/src/lib.rs:1347]: `Component::ParentDir`

**Review by mattsse — CHANGES_REQUESTED** [review]
awesome!
just tested this and works as expected.
since this is primarily etherscan related, I'd suggest we add support for this directly into ethers-etherscan::ContractMetadata instead

perhaps some like `fn into_source_tree|expand() -> Result<SourceTree>` and the sourcetree (maybe a better name?) object would essentially hold path and content of all files?

and then `SourceTree::write_to(dir: Path)` would dump them into a dir

**Comment by gakonst** [comment]
Hey @jubos thanks for taking a stab at this PR! Wonder what your plans are for getting this over the line? Happy to help with any specific code you don't have time to write

**Comment by jubos** [comment]
Planning to get it over the line this weekend.

**Comment by jubos** [comment]
I made this PR to ethers-rs: https://github.com/gakonst/ethers-rs/pull/990 that incorporates a lot of @mattsse's suggestions.  We can discuss over there and this PR will end up being very abbreviated.

**Comment by mattsse** [comment]
> I made this PR to ethers-rs: [gakonst/ethers-rs#990](https://github.com/gakonst/ethers-rs/pull/990) that incorporates a lot of @mattsse's suggestions. We can discuss over there and this PR will end up being very abbreviated.

Great! so we can drastically simplify this once merged?

**Comment by gakonst** [comment]
@jubos merged on ethers!

**Comment by jubos** [comment]
Added this PR to centralize the error handling for unverified contracts.  https://github.com/gakonst/ethers-rs/pull/1065.

Once that lands, can simplify this and other callsites in foundry.

**Review by gakonst — APPROVED** [review]

**Review by mattsse — APPROVED** [review]
lgtm

**Comment by gakonst** [comment]
This is great @jubos! Love it. 

Testing `cargo run --bin cast -- etherscan-source 0x8d04a8c79cEB0889Bdd12acdF3Fa9D207eD3Ff63 -d blitmap` it creates a `blitmap/Blitmap/` directory, wonder if instead it should be just `blitmap`?

Similarly, testing `cargo run --bin cast -- etherscan-source 0xBB9bc244D798123fDe783fCc1C72d3Bb8C189413 -d thedao` produces `thedao/DAO/DAO` (notice that there's no `.sol` extension).

If the above 2 are unintended, do you mind addressing it so that there's no double directory and that all files have a .sol extension? Otherwise we're good for merge, and would encourage you to do a twitter ann of this as a feature :)

**Comment by jubos** [comment]
Added the .sol change to ethers-rs: https://github.com/gakonst/ethers-rs/pull/1077.

We could land this commit as is and then will get this improved behavior when the ether-rs commit lands.
